RFCs/0033-Hierarchy-TestCase-Property.md

```diff
@@ -0,0 +1,29 @@
+
+
+# 0033 Hierarchy TestCase Property
+
+## Summary
+This document standardizes an additional `Hierarchy` property on TestCase to support more flexible visual display of tests in the Visual Studio Test Explorer.
+
+## Overview
+The `ManagedType` and `ManagedMethod` properties [have been defined](0017-Managed-TestCase-Properties.md) as a way of deterministically identifying the type and method to which a test case belongs (in managed code). However, the legacy `FullyQualifiedName` property is still required to deduce a Namespace, Classname and TestGroup for display in the Test Explorer hierarchy. When the Test Explorer added a hierarchy for easier grouping and navigation of tests, there was no specified way for a TestCase to provide those values so they were heuristically derived from `FullyQualifiedName`. 
+
+This document specifies a way for the TestCase to provide these display values directly to the test explorer, granting flexibility and control back to the test adapter.
+
+## Specification
+
+TestCases for managed code may include a string array (string[]) valued property named `Hierarchy`. The specification below outlines the requirements for the contents of the property.
+
+### `Hierarchy` Property
+
+The `Hierarchy` test case property represents the text strings used to display the test in the Visual Studio Test Explorer window. 
+
+* The property must be an array of strings with 4 elements. If the property is null, missing or doesn't have 4 elements then it is is ignored as invalid.
+* The element values are arranged as `[TestGroup, Class, Namespace, Container]`, which is the reverse of the order of display in the Test Explorer. i.e. Container is the top level node, Namespace is next, etc... Each TestCase is coalesced into a tree using the following order from the root. `Container`, `Namespace`, `Class`, `TestGroup`, `DisplayName`.
```

Comment:
Turning the array of strings into 4 separate custom properties would add a ton of overhead and I don't think it's worth the extra size and CPU cost. 

We have discussed adding more slots to the array in the future if needed. That would probably be the simplest way to accommodate new display parameters and would not increase the cost very much at all. We would use the size of the array to determine which "spec" the test case is written for (i.e. today the array must be length=4 or it is invalid).

Reply:
So maybe it is better to fix the underlying problem of not being able to send data efficiently, rather than introducing a workaround that won't version well?

> adding more slots to the array in the future if needed. 

How do we version that? we make the array 5 items to add something that belongs after container, then 6 items then 7 items, and output empty values as placeholders?

Reply:
I would love to fix the efficiency issues of sending custom properties, but even if we did we still have to support older protocol versions for a while and I don't want to make this feature dependent upon future updates to TP that could hinder quick rollout/adoption. That being said, I don't think there is anything particularly bad about representing the hierarchy as an array of strings.

> How do we version that? we make the array 5 items to add something that belongs after container, then 6 items then 7 items, and output empty values as placeholders?

Yes we make the array 5 items long, or 6 items long as needed. Null or empty slots don't cost much if anything. Just a note, that the order of the array has changed back so that `Container` is the first item.

Reply:
Or we can simply make them real well-known properties on the object model object, this needs support from testhost anyway, doesn't it? :)

Reply:
> Or we can simply make them real well-known properties on the object model object, this needs support from testhost anyway, doesn't it? :)

We tried to do this originally, but that was a nonstarter due to the fact that we can't rely on having a new enough version of the object model assembly available in the test host.

Reply:
I would like to know more about the exact scenario. Is that because the properties are set by adapter and adapter does not claim dependency on a specific minimum version of ObjectModel?